### PR TITLE
N+1 problem and integration test

### PR DESCRIPTION
**Pull request by anyepeter**
Hello 👋😁 CR, In this PR, we fix the N+1 problem and did the integration test for the pages with the following requirements 🔖 
### General requirements
✅ Make sure that there are [no linter errors](https://github.com/microverseinc/linters-config).
✅ Make sure that you used correct [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
✅ Make sure that you documented your work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).
### Ruby requirements
✅ Follow our list of [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).
### Project requirements
- n+1
✅ Make sure that the N+1 problem is solved when fetching all posts and their comments for a user.
✅ Do not use the bullet gem in this exercise - instead analyze the logs as showed in this [article](https://dev.to/junko911/rails-n-1-queries-and-eager-loading-10eh)
✅ In your PR description include the screenshots of your app's logs before and after the fix

#### Before
imag1

#### After
image2

- Integration specs
✅ Use Capybara to write integration tests for each view in your project.

- User index page:

✅ I can see the username of all other users.
✅ I can see the profile picture for each user.
✅ I can see the number of posts each user has written.
✅ When I click on a user, I am redirected to that user's show page.

 - user show page:
✅ I can see the user's profile picture.
✅ I can see the user's username.
✅ I can see the number of posts the user has written.
✅ I can see the user's bio.
✅ I can see the user's first 3 posts.
✅ I can see a button that lets me view all of a user's posts.
✅ When I click a user's post, it redirects me to that post's show page.
✅ When I click to see all posts, it redirects me to the user's post's index page.

- User post index page:
✅ I can see the user's profile picture.
✅ I can see the user's username.
✅ I can see the number of posts the user has written.
✅ I can see a post's title.
✅ I can see some of the post's body.
✅ I can see the first comments on a post.
✅ I can see how many comments a post has.
✅ I can see how many likes a post has.
✅ I can see a section for pagination if there are more posts than fit on the view.
✅ When I click on a post, it redirects me to that post's show page.
- Post show page:

✅ I can see the post's title.
✅ I can see who wrote the post.
✅ I can see how many comments it has.
✅ I can see how many likes it has.
✅ I can see the post body.
✅ I can see the username of each commentor.
✅ I can see the comment each commentor left.
✅ Be sure to include integration specs for any other features you may have implemented!

### CODE REVIEW

File: spec/integration/user_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+RSpec.describe 'Users page', type: :feature do
+  before(:all) do
+    visit users_path
+  end
+  describe 'display the user index page' do
+    it 'should have all usernames for all users' do
+      expect(page).to have_selector('h2', text: 'Tom')
+      expect(page).to have_selector('h2', text: 'Lilly')
+      expect(page).to have_selector('h2', text: 'Peter')
```
Comment (ichala) [inline spec/integration/user_index_spec.rb]: - [ ] as part of the requirements, kindly add these missing tests for the user index : 

- Tests that asserts you can see the username of all other users.(currently you are testing only one user)
- Tests that asserts you can see the profile picture for each user.
- Tests that asserts you can see the number of posts each user has written.
- Tests that redirect to the user's show page when you click on a user.

File: spec/integration/user_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      # navigate to the show page
+      sleep(1)
+      click_on('Tom')
+
+      # assert the displayed page
+      sleep(1)
+      assert_text 'Bio'
+    end
+  end
+end
```
Comment (ichala) [inline spec/integration/user_show_spec.rb]: - [ ] as part of the requirements, kindly add these missing tests for the user show: 

* Tests that asserts you can see the user's profile picture.
* Tests that asserts you can see the number of posts the user has written.
* Tests that asserts you can see the user's first 3 posts.
* Tests that asserts you can see a button that lets me view all of a user's posts.
* Tests that redirects to the post's show page when you click on a user's post.
* Tests that redirects to the user's post's index page when you click on a **see all posts**.

File: spec/integration/post_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+      sleep(1)
+      fill_in 'post[Title]', with: 'Yet another post..'
+      fill_in 'post[Text]', with: 'This is another post hope you like reading..'
+
+      sleep(1)
+      click_button 'Create Post'
+      sleep(2)
+    end
+  end
+end
```
Comment (ichala) [inline spec/integration/post_index_spec.rb]: - [ ]  as part of the requirements, kindly add these missing tests for the post index :


* Tests that asserts you can see the user's profile picture.
* Tests that asserts you can see the user's username.
* Tests that asserts you can see the number of posts the user has written.
* Tests that asserts you can see a post's title.
* Tests that asserts you can see some of the post's body.
* Tests that asserts you can see the first comments on a post.
* Tests that asserts you can see how many comments a post has.
* Tests that asserts you can see how many likes a post has.
* Tests that asserts you can see a section for pagination if there are more posts than fit on the view.
* Tests that redirects to the post's show page when you click on a post.

File: spec/integration/post_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+      fill_in 'comment[text]', with: 'My pleasure Tom!'
+
+      sleep(1)
+      click_button('Create Comment')
+
+      # add a like
+      sleep(1)
+      click_button('Save Like')
+    end
+  end
```
Comment (ichala) [inline spec/integration/post_show_spec.rb]: - [ ]  as part of the requirements, kindly add these missing tests for the user index :

* Tests that asserts you can see how many comments it has.
* Tests that asserts you can see how many likes it has.
* Tests that asserts you can see the post body.
* Tests that asserts you can see the username of each commentor.
* Tests that asserts you can see the comment each commentor left.

File: spec/integration/post_show_spec.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      # navigate to the posts page
+      visit user_posts_path(1)
+
+      # assert the displayed page
+      assert_text 'hardware'
+
+      sleep(1)
+      click_on 'hardware'
+
+      # page.click_link(:id => 'post_link')
```
Comment (ichala) [inline spec/integration/post_show_spec.rb]: - [ ] to keep your repo clean kindly remove any commented code from your project 🙏

**Review by ichala — CHANGES_REQUESTED** [review]
Hi Student
Good job so far! 
👏👏👏 
You still need to work on some issues to go to the next project, but you're almost there!

the-simpsons-homer-simpson

## To Highlight 🎉

- No Linter errors  ✔️ 
- used `includes` to fix n+1 problems ✔️ 


 ## Required Changes ♻️

- [ ] **_🛑 as part of the requirements  Kindly  add screenshots of the logs before and after the fix to the PR description_**
- [ ] **_🛑  Check The Comments Below_**





## Optional suggestions

- N/A

**********
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by mikethreels — APPROVED** [review]
Hi Peter, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
- linters are passing
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**